### PR TITLE
Focal CN file prep: use Gencode gtf to minimize ensembl + gene symbol mapping discrepancies

### PR DESCRIPTION
**Pull request by komalsrathi**
#### What scientific question is your analysis addressing?

1. Using [org.Hs.eg.db](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/dev/analyses/focal-cn-file-preparation/04-prepare-cn-file.R#L75-L78) to extract gene symbols causes discrepancies in mapping Ensembl gene identifiers to gene symbols. We need to change the code such that the Ensembl identifier and gene symbol both are being pulled from the input Gencode gtf file and only the cytoband information is being extracted from `org.Hs.eg.db `.

2. Reduce number of comparisons used for overlap matching between the cnv dataframe and the exon coordinates to improve efficiency. 

#### What was your approach?

1. Use Gencode gtf file as a single source to obtain the exon coordinates, ensembl gene identifier and gene symbol. Only use `org.Hs.eg.db` to pull cytoband by joining on **both** `ensembl identifier + gene symbol` to avoid discrepancies in gene id + symbol mapping. Examples of some discrepancies using the old vs new method:

*  CNVkit autosomes output

Using the old method, `ENSG00000248472` maps to `DDX11L16`
```
old <- data.table::fread('~/Downloads/cnvkit_annotated_cn_wxs_autosomes.tsv.gz')
old %>%
  filter(ensembl == "ENSG00000248472") %>%
  dplyr::select(ensembl, gene_symbol) %>%
  unique() 
           ensembl gene_symbol
1: ENSG00000248472    DDX11L16
```

Using the update method, `ENSG00000248472` maps to `DDX11L9`:
```
new <- data.table::fread('results/cnvkit_annotated_cn_wxs_autosomes.tsv.gz')
new %>%
  filter(ensembl == "ENSG00000248472") %>%
  dplyr::select(ensembl, gene_symbol) %>%
  unique()
           ensembl gene_symbol
1: ENSG00000248472     DDX11L9
```

* CNVkit sex chromosomes output

Old output, `CT47A11` maps to two Ensembl identifiers using `org.Hs.eg.db`:
```
old <- data.table::fread('~/Downloads/cnvkit_annotated_cn_wxs_x_and_y.tsv.gz')
old %>%
  filter(gene_symbol == "CT47A11") %>%
  dplyr::select(ensembl, gene_symbol) %>%
  unique() 
           ensembl gene_symbol
1: ENSG00000226929     CT47A11
2: ENSG00000226023     CT47A11
```

New output, `CT47A11` maps to one Ensembl identifier based on Gencode gtf v27:
```
new <- data.table::fread('results/cnvkit_annotated_cn_wxs_x_and_y.tsv.gz')
new %>%
  filter(gene_symbol == "CT47A11") %>%
  dplyr::select(ensembl, gene_symbol) %>%
  unique()
           ensembl gene_symbol
1: ENSG00000226929     CT47A11
```

2. For autosome calculation, remove not only `chrX` and `chrY` but also several other contigs with `_` in the names. This will reduce computation time used for matching overlaps.
3. Refactor approach for matching overlaps: First `sort` the coordinates of both CNV data frame and the gencode-based exon coordinates. Then, use breaks of 100 instead of 200 because matching is usually quicker on shorter data-frames or genomic-ranges. Next, use `IRanges::findOverlaps` to find overlaps, I found this ran pretty fast on my local machine compared to `mergeByOverlaps` but haven't benchmarked. Further, to increase performance by reducing the number of comparisons, filter the sequences (i.e. chromosomes) in the subject data-frame (i.e. Gencode exon coordinates) to only those that are present in query data-frame(i.e CNV coordinates).  

#### What GitHub issue does your pull request address?

N/A

#### Which areas should receive a particularly close look?

1. Approach of creating GenomicRanges object containing exon coordinates from Gencode gtf 
2. Approach for matching overlaps and use of `findOverlaps` within `process_annotate_overlaps` function
3. I generated the outputs (not pushed because even after using git lfs, it is exceeding the quota) by running the run script as following, which I wasn't sure if it is the correct way:  

```
bash run-prepare-cn-OpenTarget.sh
RUN_ORIGINAL=1 bash run-prepare-cn-OpenTarget.sh
```

#### Is there anything that you want to discuss further?

1. It is suggested within the `README.md` and the `04-prepare-cn-file.R` script that we are removing copy neutral segments but I don't think we are? So, we either need to edit the script to remove copy neutral segments or fix the documentation. 

```
# unique status value contains "neutral" copy segments
> unique(autosome_annotated_cn$status)
[1] "loss"          "neutral"       "gain"          NA              "deep deletion" "amplification"
```

2. If we are not using the `run-prepare-cn.sh` script, maybe we should remove it?

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes

#### What types of results are included (e.g., table, figure)?

Tables

#### What is your summary of the results?

md5sums:

```
cnvkit_annotated_cn_autosomes.tsv.gz	7b2be9b3ce6d7be9dd2ec9f4b3c7372c 
cnvkit_annotated_cn_wxs_autosomes.tsv.gz	72e72b43714ad7109db4dba79463297c 
cnvkit_annotated_cn_wxs_x_and_y.tsv.gz	a7d2ab56907e51cd55d34ffa4422805e 
cnvkit_annotated_cn_x_and_y.tsv.gz	ca024a823bd2816643f66437f6e7d144 
consensus_seg_annotated_cn_autosomes.tsv.gz	3abf523ff84c28a7d756d6e68f50a168 
consensus_seg_annotated_cn_x_and_y.tsv.gz	116cf1f678b8dca86adabc24f54c8730 
consensus_seg_most_focal_cn_status.tsv.gz	1d81ad1cf601c949f0c616adb92e1ca0 
consensus_seg_with_ucsc_cytoband_status.tsv.gz	2d8990a782d3e226b7f0ea66b8128dbe 
consensus_wgs_plus_cnvkit_wxs.tsv.gz	3eee6248ec625574be008f05d0baf9b9 
consensus_wgs_plus_cnvkit_wxs_autosomes.tsv.gz	95b633794bef98f43d4f19b9d3dfa039 
consensus_wgs_plus_cnvkit_wxs_x_and_y.tsv.gz	0b8c8ad78b3950f7e6837bf3cb92804e 
controlfreec_annotated_cn_autosomes.tsv.gz	6d443819228658c0a768e236aa2715ec 
controlfreec_annotated_cn_wxs_autosomes.tsv.gz	8af21a2a559b6f55eec9c6a43bc24148 
controlfreec_annotated_cn_wxs_x_and_y.tsv.gz	81e1bc17e9890c63f555931d8610e21d 
controlfreec_annotated_cn_x_and_y.tsv.gz	70a8e9b09b5b0d4670e8146d612f8dbe 
```

number of lines per file (includes header line):
```
cnvkit_annotated_cn_autosomes.tsv.gz	7810672
cnvkit_annotated_cn_wxs_autosomes.tsv.gz	26912446
cnvkit_annotated_cn_wxs_x_and_y.tsv.gz	1337194
cnvkit_annotated_cn_x_and_y.tsv.gz	748761
consensus_seg_annotated_cn_autosomes.tsv.gz	891899
consensus_seg_annotated_cn_x_and_y.tsv.gz	61044
consensus_seg_most_focal_cn_status.tsv.gz	750307
consensus_seg_with_ucsc_cytoband_status.tsv.gz	624089
consensus_wgs_plus_cnvkit_wxs_autosomes.tsv.gz	27804344
consensus_wgs_plus_cnvkit_wxs_x_and_y.tsv.gz	1398237
consensus_wgs_plus_cnvkit_wxs.tsv.gz	29202580
controlfreec_annotated_cn_autosomes.tsv.gz	3085603
controlfreec_annotated_cn_wxs_autosomes.tsv.gz	6891861
controlfreec_annotated_cn_wxs_x_and_y.tsv.gz	518224
controlfreec_annotated_cn_x_and_y.tsv.gz	184909
```

#### Reproducibility Checklist

- [ ] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.

#### Documentation Checklist

- [ ] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by komalsrathi** [comment]
@runjin326 I wasn't sure who else to add but feel free to add more people as reviewers if needed. Also please note (as mentioned in the description) that I ran this the following scripts on my local machine and generated the outputs but they were not pushed because even after using `git lfs`, it is exceeding the quota of this repo.  

```
bash run-prepare-cn-OpenTarget.sh
RUN_ORIGINAL=1 bash run-prepare-cn-OpenTarget.sh
```

**Comment by runjin326** [comment]
@komalsrathi , thanks so much for preparing this! Maybe you can post the md5sum for the results? or a `wc -l ` call. 
Two questions to confirm: 
1) the nodule is ran with v10 data release?
2) did you run on the OT docker?

**Comment by komalsrathi** [comment]
> @komalsrathi , thanks so much for preparing this! Maybe you can post the md5sum for the results? or a `wc -l ` call. Two questions to confirm:
> 
> 1. the nodule is ran with v10 data release?
> 2. did you run on the OT docker?

Yes I ran with v10 and I did not use the docker instead just ran via command line bash. I added the md5sums + number of lines in the summary section of the description.

File: analyses/focal-cn-file-preparation/04-prepare-cn-file.R [hunk 1]
```diff
@@ -263,44 +175,45 @@ for (i in 1:length(slice_vector)){
...
+if(runWXSonly){
+  autosome_output_file <- paste0(filename_lead, "_wxs_autosomes.tsv.gz")
+} else {
+  autosome_output_file <- paste0(filename_lead, "_autosomes.tsv.gz")
 }
 
 # Save final data.frame to a tsv file
-readr::write_tsv(autosome_annotated_cn,
-                 file.path(results_dir, autosome_output_file))
+readr::write_tsv(x = autosome_annotated_cn, file = file.path(results_dir, autosome_output_file))
```
Comment (runjin326) [inline analyses/focal-cn-file-preparation/04-prepare-cn-file.R]:
```suggestion
readr::write_tsv(autosome_annotated_cn, file.path(results_dir, autosome_output_file))
```
when running on the Docker, it throws an error message 
```
Error in readr::write_tsv(x = autosome_annotated_cn, file = file.path(results_dir,  : 
  unused argument (file = file.path(results_dir, autosome_output_file))
``` 
- modifying to original format works for me.

File: analyses/focal-cn-file-preparation/04-prepare-cn-file.R [hunk 2]
```diff
@@ -339,18 +252,18 @@ if (xy_flag) {
...
+  if(runWXSonly){
+    sex_chrom_output_file <- paste0(filename_lead, "_wxs_x_and_y.tsv.gz")
+  } else {
+    sex_chrom_output_file <- paste0(filename_lead, "_x_and_y.tsv.gz")
   }
   
   # Save final data.frame to a tsv file
-  readr::write_tsv(sex_chrom_annotated_cn,
-                   file.path(results_dir, sex_chrom_output_file))
+  readr::write_tsv(x = sex_chrom_annotated_cn, file = file.path(results_dir, sex_chrom_output_file))
```
Comment (runjin326) [inline analyses/focal-cn-file-preparation/04-prepare-cn-file.R]:
```suggestion
  readr::write_tsv(sex_chrom_annotated_cn, file.path(results_dir, sex_chrom_output_file))
```

File: analyses/focal-cn-file-preparation/04-prepare-cn-file.R (hunk 1 above)
Comment (komalsrathi) [reply analyses/focal-cn-file-preparation/04-prepare-cn-file.R]: That is super weird, I am pretty sure x and file are valid parameters for write_tsv. I guess the Dockerfile must be using a very old version of readr.

**Review by runjin326 — CHANGES_REQUESTED** [review]
@komalsrathi , thanks so much for preparing this! 
Additional to the write file error - when I was looking at the result file `consensus_wgs_plus_cnvkit_wxs_x_and_y.tsv.gz`, the X and Y annotation is a bit weird - seemed like same transcript + gene symbol can match to different cytoband. Could you please look into this? Also, the old file does not have cytoband that is just X or Y so am curious how that came about. 
image
For the old file, for example `AKAP17A` matches to `Xp22.33` only. 

I think a more informative QC of the results would be to compare the re-run results to data release and compare the results to make sure that all the differences that we see are from ensemble + gene symbol match different, and each sample still have the same status + cytoband in the file ? Could you please look into that as well?

**Comment by komalsrathi** [comment]
For your comment regarding multiple mapping, that is how it is when you pull from `org.Hs.eg.db`:

This query:
```
AnnotationDbi::select(org.Hs.eg.db::org.Hs.eg.db, # database
                      keys = unique('ENSG00000185291'),  # data to use for retrieval
                      columns = c("MAP","ENSEMBL","SYMBOL"), # information to retrieve for given data
                      keytype = "ENSEMBL")
```

will return:
```
          ENSEMBL MAP SYMBOL
1 ENSG00000185291   X  IL3RA
2 ENSG00000185291   Y  IL3RA
```

The old file is pretty much not accurate because the cytoband info is being pulled only using Ensembl gene identifier rather it should take BOTH gene symbol AND gene identifier to pull the cytoband info.

**Comment by komalsrathi** [comment]
For `AKAP17A`, I get X and Y as well: 

```
AnnotationDbi::select(org.Hs.eg.db::org.Hs.eg.db, # database
                      keys = unique("AKAP17A"),  # data to use for retrieval
                      columns = c("MAP","ENSEMBL","SYMBOL"), # information to retrieve for given data
                      keytype = "SYMBOL")

   SYMBOL MAP         ENSEMBL
1 AKAP17A   X ENSG00000197976
2 AKAP17A   Y ENSG00000197976
```

I am using org.Hs.eg.db_3.13.0 so it could be because of the package version differences? Unsure of the version being used in OpenPedCan, can you let me know so I can rerun using that? On a side note, for annotation purposes,  shouldn't we always use the latest version?

**Comment by runjin326** [comment]
@komalsrathi, I am not exactly sure what org.Hs.eg.db version it is running for the v10 release - maybe @Bo will know since the last couple times it was generated in CAVATICA? 
But I am looking into those genes, and looks like there are some ambiguity as to where they map to. And looks like Ensembl calls it `Xp22.33` and Entrez Gene calls `X;Y` 
image
Interestingly, I checked the output when running in the docker and the results are like this:
```
> mine %>% filter(gene_symbol == "AKAP17A")
# A tibble: 2,492 × 8
   biospecimen_id           status        copy_number ploidy ensembl         gene_symbol cytoband germline_sex_es…
   <chr>                    <chr>               <dbl>  <dbl> <chr>           <chr>       <chr>    <chr>           
 1 TARGET-10-PANVUD-09A-01D neutral                 2      2 ENSG00000197976 AKAP17A     Xp22.33  NA              
 2 TARGET-10-PANVUD-09A-01D neutral                 2      2 ENSG00000197976 AKAP17A     Yp11.2   NA              
 3 BS_AEEKJ1QK              deep deletion           0     NA ENSG00000197976 AKAP17A     Xp22.33  Male            
 4 BS_AEEKJ1QK              deep deletion           0     NA ENSG00000197976 AKAP17A     Yp11.2   Male            
 5 BS_WYTDVC0Y              NA                      1     NA ENSG00000197976 AKAP17A     Xp22.33  Male            
 6 BS_WYTDVC0Y              NA                      1     NA ENSG00000197976 AKAP17A     Yp11.2   Male            
 7 TARGET-10-PANVGZ-09A-01D neutral                 1      2 ENSG00000197976 AKAP17A     Xp22.33  NA              
 8 TARGET-10-PANVGZ-09A-01D neutral                 1      2 ENSG00000197976 AKAP17A     Yp11.2   NA              
 9 TARGET-10-PANTBB-09A-01D loss                    1      3 ENSG00000197976 AKAP17A     Xp22.33  NA              
10 TARGET-10-PANTBB-09A-01D loss                    1      3 ENSG00000197976 AKAP17A     Yp11.2   NA          
```
version for the old results maps only to `Xp22.33` 

I am guessing CAVATICA, Docker, and your local machine are using different versions.

I think the next steps can be:
1) Could you please re-run it using the docker image from OT and see whether your results are the same as mine? 
2) Open a ticket and work with Bo to synchronize the version of `org.Hs.eg.db` so that the results are comparable

The code change itself makes sense and works well for me otherwise.

**Comment by komalsrathi** [comment]
Based on our slack discussion, pulling in Bo. @zhangb1 can you tell me what version of `org.Hs.eg.db` and `AnnotationDbi_1.54.1` R packages you are using on CAVATICA? Seems like I m getting completely different output using version `org.Hs.eg.db_3.13.0`

**Comment by zhangb1** [comment]
I am using the file: `txdb_from_gencode.v27.gtf.db`

I use the verison `3.10.0` for org.Hs.eg.db  , we have a docker file : https://github.com/d3b-center/d3b-bixu-data-assembly/blob/master/docker/Dockerfile

**Comment by komalsrathi** [comment]
@runjin326 I tried running this via docker (thanks for the directions) but unfortunately it is being killed even after slicing the df into chunks. The weird part is without docker and with the package versions that I have installed on my machine this module ran very quick on my local machine without any memory issues. I am not sure how else to test this out.

File: analyses/focal-cn-file-preparation/04-prepare-cn-file.R (hunk 1 above)
Comment (komalsrathi) [reply analyses/focal-cn-file-preparation/04-prepare-cn-file.R]: I just committed this change - thanks. Like you said, it wasn't working on the docker image.

**Comment by komalsrathi** [comment]
I just re-ran (again thanks to @runjin326 for the directions reg. changing docker resources) and updated the md5sums + number of lines. 

Both of the above issues seem to be addressed:

```
# AKAP17A still maps to both X and Y after re-running 
➜  gzcat consensus_wgs_plus_cnvkit_wxs_x_and_y.tsv.gz | grep 'AKAP17A' | cut -f7 | sort | uniq
Xp22.33
Yp11.2

# issue with X/Y seems to have been resolved
➜ gzcat consensus_wgs_plus_cnvkit_wxs_x_and_y.tsv.gz | grep IL3RA | cut -f7 | sort | uniq    
Xp22.33
Yp11.2
```

**Review by runjin326 — APPROVED** [review]
@komalsrathi , thanks so much for spotting this and preparing the PR - my re-ran got the exact same results as yours and all the logic and results make sense to me!

**Comment by runjin326** [comment]
@ewafula, could you please also review this PR whenever you get a chance? Thanks!

**Comment by ewafula** [comment]
> @ewafula, could you please also review this PR whenever you get a chance? Thanks!

Sure, I will review.

**Review by ewafula — APPROVED** [review]
Thanks, @komalsrathi for working on the PR. The code update and the logic makes sense. I obtained the exact same results as yours including `md5sum` and the `number of lines per file` by running the update PR in OpenPedCan Docker. I also compared the `consensus_wgs_plus_cnvkit_wxs_autosomes.tsv.gz`, `consensus_wgs_plus_cnvkit_wxs.tsv.gz`, and `consensus_wgs_plus_cnvkit_wxs_x_and_y.tsv.gz` result files with those in the OpenPedCan v10 data release. There are differences in `Ensembl` to `gene_symbol`mappings. For example, in `consensus_wgs_plus_cnvkit_wxs_autosomes.tsv.gz` there are 10,237 unique mappings present in v10 data, not in the PR results, and 171 unique mappings in PR results and not in v10 data. I am assuming these are discrepancies being addressed by the PR. 
Approving!
Cc @runjin326

**Comment by runjin326** [comment]
@ewafula, thanks Eric! And yes, the modification for this module is for getting an unambiguous `Ensembl` to `gene_symbol` mappings so the difference that you see is expected.
@komalsrathi, could you please merge the most up-to-date dev to this branch? I will merge this PR after that. Thanks!

**Comment by komalsrathi** [comment]
@runjin326 sure, done! thanks!

**Comment by runjin326** [comment]
@komalsrathi, weird- looks like the docker check did not pass? have not seen this one before - any idea why?

**Comment by komalsrathi** [comment]
> @komalsrathi, weird- looks like the docker check did not pass? have not seen this one before - any idea why?

same, I am not sure about this one. It looks like the error is related to:

```
Error: buildx failed with: error: failed to solve: process "/bin/sh -c wget https://github.com/bedops/bedops/releases/download/v2.4.37/bedops_linux_x86_64-v2.4.37.tar.bz2 &&     tar -jxvf bedops_linux_x86_64-v2.4.37.tar.bz2 &&     rm -f bedops_linux_x86_64-v2.4.37.tar.bz2 &&     mv bin/* /usr/local/bin" did not complete successfully: exit code: 8
```

weird why this only showed up now as it is unrelated to the PR...

**Comment by runjin326** [comment]
> weird why this only showed up now as it is unrelated to the PR...

 Exactly - maybe you can make a change such as add a line somewhere and then commit and push again - and let the check run again to see whether it fixes it? Sometimes it fixes stuff for me in OpenPBTA (not sure why though).

**Comment by logstar** [comment]
@komalsrathi @runjin326 I agree. We could just rerun with the same file to see whether the docker check passes.

The docker check error is probably caused by a GitHub problem, which is also reported at <https://github.com/github/feedback/discussions/8535>.

Attached is the screenshot of the docker check error.

Screen Shot 2021-12-08 at 10 38 22 AM

**Comment by komalsrathi** [comment]
> > weird why this only showed up now as it is unrelated to the PR...
> 
> Exactly - maybe you can make a change such as add a line somewhere and then commit and push again - and let the check run again to see whether it fixes it? Sometimes it fixes stuff for me in OpenPBTA (not sure why though).

That worked!

**Pull request merged**